src/vsphere_cpi/lib/cloud/vsphere/selection_pipeline.rb

```diff
@@ -0,0 +1,125 @@
+module VSphereCloud
+  class SelectionPipeline
+    include Enumerable
+    include Logger
+
+    module StableRandom
+      def self.[](object)
+        @memo ||= {}
+        @memo[object] ||= Random.rand
+      end
+    end
+
+    # @return [Object] the object to select placements for
+    attr_reader :object
+
+    def self.with_filter(*args, &block)
+      filter_list << { args: args, block: block }
+    end
+
+    def self.filter_list
+      @filter_list ||= []
+    end
+
+    def self.with_scorer(*args, &block)
+      scorer_list << { args: args, block: block }
+    end
+
+    def self.scorer_list
+      @scorer_list ||= []
+    end
+
+    # @param object [Object] the object to select placements for
+    def initialize(object, *args)
+      raise ArgumentError, 'No gather block provided' unless block_given?
+
+      @object = object
+      @gather = Proc.new
+
+      self.class.filter_list.each do |filter|
+        with_filter(*filter[:args], &filter[:block])
+      end
+
+      self.class.scorer_list.each do |scorer|
+        with_scorer(*scorer[:args], &scorer[:block])
+      end
+    end
+
+    def inspect
+      "#<SelectionPipeline @filter_list #{filter_list} @scorer_list #{scorer_list}>"
+    end
+
+    def accept?(placement)
+      result = filter_list.all? do |filter|
+        result = filter.call(placement, @object)
+        logger.info("Rejecting #{placement.inspect}") unless result
```

Comment:
here as well, we should add based on what filter a placement was rejected will be more informative